### ISSUE TITLE
Add infrastructure to isolate the contents of the Advice Provider based on the call context.

### ISSUE THREAD

**Issue by hackaugusto** [0xPolygonMiden/miden-vm#1210]
@phklive asked me about this, and I couldn't find a way to separate the state in the advice provider based on the current call context.

This can be an issue for the tx kernel, as it is, it may be possible for a note to inject data into the advice provider, either adding new data or overwriting data, that later on be used by the tx kernel or the account code. This could have unintended side effects, so it may be best to isolate the state of the advice provider depending on the call context.

**Comment by bitwalker** [comment]
@bobbinth @plafer @huitseeker Is this still relevant/desirable? AFAIK, we can't overwrite entries in the advice map, and by the time the transaction kernel calls into user code, the contents of the advice stack wouldn't matter any more - not to mention we're supposed to be validating everything we get from the advice provider anyway.

**Comment by plafer** [comment]
Indeed, injection of malicious data by a note shouldn't be an issue, since as you mentioned we always validate any advice provider data. Thinking about this, a _potential_ concern is if a call context writes sensitive information to the advice map, which can then later be read by another call context (and namely if some user code can read sensitive data written by the tx kernel). Though it's probably not an issue in practice, as I wouldn't expect the tx kernel to write anything sensitive to the advice map.

**Comment by bobbinth** [comment]
I'm not sure how much this is an issue and what we can do differently (at least not without introducing considerable extra complexity), especially because we use the advice provider to facilitate data transfers between calling contexts, but to clarify a few things:

- AFAIK, we can overwrite entries in the advice map. I don't think we use this in practice anywhere, and I'm not sure imposing this restriction would change things materially.
- The advice data should always be treated as untrusted, and so, the consequences of somehow injecting malicious data into the advice provider are somewhat limited:
  - This can error out execution (or maybe crush it if some code is not robust enough).
  - This can let some code read some data it is not intended to read.
- Expanding the last point, the advice provider could contain private data (e.g., account balances) - but should not contain critically-sensitive data (e.g., private key). So, if someone executes "untrusted code" in their transaction, this could would be able to access this data - but doing anything useful with this data would not be easy. The data would be gone after transaction execution, and the code would not be able to create any long-lasting artifacts. One exception to this is creating notes, but this should be detected by user's wallet software (e.g., if an unexpected note is created during a transaction, this should be made very obvious to the user).

**Comment by plafer** [comment]
Closing as not planned
